collector/pg_locks.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const locksSubsystem = "locks"
+
+func init() {
+	registerCollector(locksSubsystem, defaultEnabled, NewPGLocksCollector)
+}
+
+type PGLocksCollector struct {
+	log log.Logger
+}
+
+func NewPGLocksCollector(config collectorConfig) (Collector, error) {
+	return &PGLocksCollector{
+		log: config.logger,
+	}, nil
+}
+
+var (
+	pgLocksDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(
+			namespace,
+			locksSubsystem,
+			"count",
+		),
+		"Number of locks",
+		[]string{"datname", "mode"}, nil,
+	)
+
+	pgLocksQuery = `
+		SELECT 
+		  pg_database.datname as datname,
+		  tmp.mode as mode,
+		  COALESCE(count, 0) as count 
+		FROM 
+		  (
+		    VALUES 
+		      ('accesssharelock'), 
+		      ('rowsharelock'), 
+		      ('rowexclusivelock'), 
+		      ('shareupdateexclusivelock'), 
+		      ('sharelock'), 
+		      ('sharerowexclusivelock'), 
+		      ('exclusivelock'), 
+		      ('accessexclusivelock'), 
+		      ('sireadlock')
+		  ) AS tmp(mode) CROSS 
```

Comment:
Just a preference, but I think I like `CROSS` on the next line.

Reply:
Yea, I don't have a preference on SQL formatters. Maybe we should pick and document a recommended one for the codebase.